File: Top_Interview_150/Array_String/python/135.py
from typing import List


class Solution:
    def candy(self, ratings: List[int]) -> int:

        n = len(ratings)
        if n == 1:
            return 1

        dp = [1] * n
        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                dp[i] = dp[i - 1] + 1

        count = dp[i]
        for i in range(n - 1, 0, -1):
            if ratings[i] < ratings[i - 1]:
                dp[i - 1] = max(dp[i - 1], dp[i] + 1)
            count += dp[i - 1]

        return count


solution = Solution()
print(solution.candy(ratings=[1, 0, 2]))

File: Top_Interview_150/Array_String/cpp/135.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int candy(vector<int> &ratings)
    {
        int n = ratings.size();
        if (n == 1)
            return 1;

        vector<int> dp(n, 1);
        for (int i = 1; i < n; ++i)
        {
            if (ratings[i] > ratings[i - 1])
                dp[i] = dp[i - 1] + 1;
        }

        int count = dp[n - 1];
        for (int i = n - 1; i > 0; --i)
        {
            if (ratings[i] < ratings[i - 1])
                dp[i - 1] = max(dp[i - 1], dp[i] + 1);
            count += dp[i - 1];
        }

        return count;
    }
};